[CALCITE-3335] ElasticSearch adapter. Introduce configuration parameter "hosts"

shikhasomani opened the pull request:
… i.e. to allow connecting on HTTPS

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -62,8 +62,9 @@ public ElasticsearchSchemaFactory() {
       final Map<String, Integer> coordinates = mapper.readValue(coordinatesString,
           new TypeReference<Map<String, Integer>>() { });
 
+      final String scheme = (String) map.get("scheme");
```
Why scheme has to be hardcoded in configuration ? 
Can it be auto-detected programmatically ?

shikhasomani replied:
Scheme can't be supplied in coordinates like:
 "coordinates": "{'https://vpc.xxx': 443}" otherwise it fails in later steps of ElasticsearchTransport. So the scheme has to be separately supplied.

asereda-gs replied:
Now when I think, it seems better to introduce another config variable like `hosts` or `nodes` which will be a JSON array of URIs:
```
  hosts: ["http://127.0.0.1:443", "http://myNode:80"]
```

shikhasomani replied:
Introduced new variable *hosts* in parallel to coordinates to maintain backward compatibility.

Either hosts or coordinates is required to connect to ES. If hosts is provided it should be a well-formed URL having protocol, host and port information. Hosts can take any number of hosts.
With coordinates it works as before.

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -89,6 +101,25 @@ private static RestClient connect(Map<String, Integer> coordinates) {
+    final Set<HttpHost> set = new LinkedHashSet<>();
+
+    for (String host: hosts) {
+      URL hostURL = new URL(host);
```
Don't use URL rather [HttpHost.create()](https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/HttpHost.html#create(java.lang.String))

shikhasomani replied:
Done

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -56,14 +59,23 @@ public ElasticsearchSchemaFactory() {
+      final RestClient client;
+
+      if (coordinatesString != null) {
+        final Map<String, Integer> coordinates = mapper.readValue(coordinatesString,
```
Are there 2 versions of `connect` function  ? 
I would rather create `List<String>` (or `List<HttpHost>`) here and pass it to unique `connect` method.
Don't forget to use [HttpHost.create()](https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/HttpHost.html#create(java.lang.String)) instead of URL

shikhasomani replied:
Done

asereda-gs commented:
Pls squash commits into one and add your name to commit message as `(FirstName LastName)`

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -56,14 +59,23 @@ public ElasticsearchSchemaFactory() {
 
-      final Map<String, Integer> coordinates = mapper.readValue(coordinatesString,
-          new TypeReference<Map<String, Integer>>() { });
+      final String hostsString = (String) map.get("hosts");
```
I prefer to composite this check like this:
```java
final List<String> hosts;
if (map.containsKey("coordinates")) {
  hosts = ...; // // build List<String> from coordinates
} else if (map.containsKey("hosts")) {
  hosts = ...; // build List<String> from hosts
} else {
   throw new IllegalArgumentException("Both 'coordinates' and 'hosts' is missing in configuration. Provide one of them.");
}
```

shikhasomani replied:
Done

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -56,15 +58,30 @@ public ElasticsearchSchemaFactory() {
 
-      // create client
-      final RestClient client = connect(coordinates);
+      final Set<HttpHost> hostSet = new LinkedHashSet<>();
```
I would rather let elastic API handle uniqueness (or not) of hosts. 
Can you declare this variable as `List<HttpHost>`  ?

asereda-gs replied:
Also please rename this variable to `hosts`

shikhasomani replied:
Done

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -56,15 +58,30 @@ public ElasticsearchSchemaFactory() {
+        final Map<String, Integer> coordinates = mapper.readValue(coordinatesString,
+                new TypeReference<Map<String, Integer>>() { });
+
+        coordinates.forEach((hostName, port) ->  hostSet.add(new HttpHost(hostName, port)));
```
instead of modifying the state (`hosts`) variable (using `forEach`) can you use streams ? 
```java
hosts = coordinates.entrySet()
.stream(entry -> new HttpHost(entry.getKey(), entry.getValue())).collect(Collectors.toList());
```

shikhasomani replied:
Done

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -75,18 +92,15 @@ public ElasticsearchSchemaFactory() {
-    for (Map.Entry<String, Integer> entry: coordinates.entrySet()) {
-      set.add(new HttpHost(entry.getKey(), entry.getValue()));
-    }
+  private static RestClient connect(Set<HttpHost> hostSet) throws MalformedURLException {
```
rename method argument : `hostSet` -> `hosts`

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -75,18 +92,15 @@ public ElasticsearchSchemaFactory() {
-    for (Map.Entry<String, Integer> entry: coordinates.entrySet()) {
-      set.add(new HttpHost(entry.getKey(), entry.getValue()));
-    }
+  private static RestClient connect(Set<HttpHost> hostSet) throws MalformedURLException {
```
Make hosts `Collection` or `List` (ES client will do the de-dupe)

shikhasomani replied:
Done.

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -56,15 +58,30 @@ public ElasticsearchSchemaFactory() {
+      final Set<HttpHost> hostSet = new LinkedHashSet<>();
+
+      if (coordinatesString != null) {
+        final Map<String, Integer> coordinates = mapper.readValue(coordinatesString,
```
For coordinates can you log a `WARNING` message  that `hosts` configuration attribute is preferred and `coordinates` is deprecated ?

asereda-gs replied:
We also need to mention it in release notes for next release (probably 1.22)

shikhasomani replied:
Added warning message.

Since this change is backward compatible it should be pushed in previous releases also till release-1.19 so it will support production environment where most of ES is on HTTPS.

asereda-gs replied:
We don't usually do back-porting. Next calcite release 1.22 should happen by end of this year

asereda-gs commented:
Pls squash your commit into one. Message should say
> [CALCITE-1234] Description (FirstName LastName)

asereda-gs commented on elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchSchemaFactory.java:
```diff
@@ -56,15 +61,38 @@ public ElasticsearchSchemaFactory() {
-      final RestClient client = connect(coordinates);
+      List<HttpHost> hosts;
+
+      if (coordinatesString != null) {
```
Prefer not to use too many temp variables. I think it can be simplified to: 
```java
if (map.containsKey("hosts")) {
 hosts =   mapper.convertValue(map.get("hosts", new TypeReference<List<String>>() { })).stream() ... ;
} else if (map.containsKey("coordinates")) {
 hosts =   mapper.convertValue(map.get("coordinates", new TypeReference<Map<String, Integer>>() { })).stream() ... ;

}

shikhasomani replied:
Reduced two variables.

shikhasomani commented:
> Pls squash commits into one and add your name to commit message as `(FirstName LastName)`

Squashed all commits and added message in the required format.

asereda-gs commented:
Pls read [contributing doc](https://calcite.apache.org/develop/#contributing). JIRA id should be in square brackets : `[CALCITE-3335]`

shikhasomani commented:
Done
